waku/waku_store_legacy/protocol.nim

```diff
@@ -44,77 +44,103 @@ type WakuStore* = ref object of LPProtocol
+    return
+
+  let reqRpc = decodeRes.value
+
+  if reqRpc.query.isNone():
+    error "empty query rpc", peerId = requestor, requestId = reqRpc.requestId
+    waku_legacy_store_errors.inc(labelValues = [emptyRpcQueryFailure])
+    # TODO: Return (BAD_REQUEST, cause: "empty query")
+    return
+
+  let
+    requestId = reqRpc.requestId
+    request = reqRpc.query.get().toAPI()
+
+  info "received history query",
+    peerId = requestor, requestId = requestId, query = request
+  waku_legacy_store_queries.inc()
+
+  var responseRes: HistoryResult
+  try:
+    responseRes = await self.queryHandler(request)
+  except Exception:
+    error "history query failed",
+      peerId = requestor, requestId = requestId, error = getCurrentExceptionMsg()
+
+    let error = HistoryError(kind: HistoryErrorKind.UNKNOWN).toRPC()
+    let response = HistoryResponseRPC(error: error)
+    return HistoryRPC(requestId: requestId, response: some(response)).encode().buffer
+
+  if responseRes.isErr():
+    error "history query failed",
+      peerId = requestor, requestId = requestId, error = responseRes.error
 
-    let
-      requestId = reqRpc.requestId
-      request = reqRpc.query.get().toAPI()
-
-    info "received history query",
-      peerId = conn.peerId, requestId = requestId, query = request
-    waku_legacy_store_queries.inc()
+    let response = responseRes.toRPC()
+    return HistoryRPC(requestId: requestId, response: some(response)).encode().buffer
 
-    var responseRes: HistoryResult
-    try:
-      responseRes = await ws.queryHandler(request)
-    except Exception:
-      error "history query failed",
-        peerId = $conn.peerId, requestId = requestId, error = getCurrentExceptionMsg()
+  let response = responseRes.toRPC()
 
-      let error = HistoryError(kind: HistoryErrorKind.UNKNOWN).toRPC()
-      let response = HistoryResponseRPC(error: error)
-      let rpc = HistoryRPC(requestId: requestId, response: some(response))
-      await conn.writeLp(rpc.encode().buffer)
-      return
+  info "sending history response",
+    peerId = requestor, requestId = requestId, messages = response.messages.len
 
-    if responseRes.isErr():
-      error "history query failed",
-        peerId = $conn.peerId, requestId = requestId, error = responseRes.error
+  return HistoryRPC(requestId: requestId, response: some(response)).encode().buffer
 
-      let response = responseRes.toRPC()
-      let rpc = HistoryRPC(requestId: requestId, response: some(response))
-      await conn.writeLp(rpc.encode().buffer)
+proc initProtocolHandler(ws: WakuStore) =
+  let rejectResponseBuf = HistoryRPC(
+    ## We will not copy and decode RPC buffer from stream only for requestId
+    ## in reject case as it is comparably too expensive and opens possible
+    ## attack surface
+    requestId: "N/A",
+    response: some(
+      HistoryResponseRPC(
+        error: HistoryError(kind: HistoryErrorKind.TOO_MANY_REQUESTS).toRPC()
+      )
```

Comment:
An analogue to previous comments. It would be interesting to have a better detail about why the limit is exceeded, i.e. what is the current rate and what it the rate-limit.

Reply:
I think this detail should not be expose to the outer world.